% v0.2
\chapter{Biological background}
In this chapter we will present basic biological terms, which we will use later in this work.
Firstly, we will describe molecules performing functions in living organisms and their representation in bioinformatics and computational biology.
Then, we will explain relationships between these molecules.
We will also provide detailed description of bacteriophages, which will be the central topic of our work.
Description of what they are and where we can find them will be provided.
Furthermore, we will explain their life cycle and we will point out some of their use.
We will also show currently valid taxonomical classification and the structure of a typical representative of this group of organisms.
At the end of this chapter we will point out what was already done in the field of their hosts prediction and what we would like to achieve.

\section{Biological molecules}
Molecules are basic components of every living organism on Earth.
Vast majority of biological function are mediated by them.
Next, we will present the central types of molecules in molecular biology, which we will use in our analysis.

\subsection{Deoxyribonucleic acid}
\emph{Deoxyribonucleic acid} (DNA) is a long chain of nucleotides (bases) of four types - adenine, cytosine, guanine and thymine.
These bases are connected through a phosphodiester bond, connection between the 3' carbon atom of one deoxyribose and the 5' carbon atom of the second deoxyribose.
The structure of DNA consists of two complementary strands coiled around each other in the form of a double helix.
In this double helix adenine is paired with thymine and cytosine with guanine through hydrogen bonds \cite{molbio}.

In bioinformatics and computational biology, we usually represent a base by its first letter - A for adenine, C for cytosine, G for guanine and T for thymine.
Due to linear organisation of nucleotides, a DNA molecule can be represented as a word from alphabet \{A, C, G, T\}, called DNA sequence. We usually store biological sequences in simple plain text files.
Probably the most commonly used format is the FASTA format.
It can consist of one or multiple records, each representing a single DNA sequence.
Each record starts with the identifier of the sequence, which is followed by lines of the DNA sequence written in direction from 5' to 3'.

\subsection{Ribonucleic acid}
\emph{Ribonucleic acid} (RNA) is, similar to DNA, it is a polymeric molecule consisting of four types of nucleotides.
There are three main differences between DNA and RNA molecules.
Firstly, the sugar-phosphate backbone of RNA contains ribose instead of deoxyribose.
This change in structure makes RNA less stable as it is more prone to hydrolysis.
Another distinctive feature of RNA is that it contains uracil instead of thymine.
Additionally, RNA appears in nature mostly as a single-stranded molecule, whereas DNA is mostly double-stranded \cite{molbio}.
This characteristic allows RNA to form more complex structures and show enzymatic activity.
In bioinformatics, sequences of RNA are usually written, similarly as with DNA, from 5' to 3', with the difference of U for uracil instead of T for thymine.

\subsection{Protein}
\emph{Proteins} are polymers consisting of amino acids connected by a peptide bond.
There are 20 basic proteinogenic amino acids encoded in DNA sequences and 2 proteinogenic amino acids incorporated into proteins by a unique mechanism.
Protein performs overwhelming majority of functions in living organisms.
Thanks to this characteristic, they will be of high importance in our analysis.
In bioinformatics, protein sequences are represented as a string of 1-letter abbreviations of their amino acids. All proteinogenic amino acids with their 1-letter codes can be found in the Table (\ref{tab:amino})

\begin{table}
 \centering
        \begin{tabular}{ l  l  l  l }
         \hline
         amino acid & 1-letter code & amino acid & 1-letter code \\
         \hline  
         alanine & A & arginine & R \\
         asparagine & N & aspartic acid & D \\
         cysteine & C & glutamine & Q \\
         glutamic acid & E & glycine & G \\
         histidine & H & isoleucine & I \\
         leucine & L & lysine & K \\
         methionine & M & phenylalanine & F \\
         proline & P & serine & S \\
         threonine & T & tryptophan & W \\
         tyrosine & Y & valine & V \\
         selenocysteine & U & pyrrolysine & O \\
         \hline
        \end{tabular}
        \caption{Amino acids}
        \label{tab:amino}
\end{table}

% upravene start
\section{Flow of genetic information}
\emph{Genome} is a complete genetic material of an organism.
It usually consists of DNA molecules, but some organisms can have genome in the form of RNA molecules.
In genome, there is written entire information about the composition of particular organism with functional elements included.
These functional elements, also called \emph{genes}, are encoded in the sequence of DNA or RNA.
Genes can be located in genomes using bioinformatics tools, which usually search for coding DNA sequences (CDSs).
CDS is a region of genome that starts with the start codon and end with the stop codon.
These regions can be directly translated into amino acid chains using \emph{standard codon table}.
In this table, every combination of three nucleotides corresponds to amino acid, with exception of the combinations TAA, TAG and TGA, which correspond to stop codons.
The relationships between biological molecules and the flow of genetic information in most living organisms are described in \emph{central dogma of molecular biology}.
Most organisms store their genetic information in the nucleus in the form of DNA.
Some parts of that DNA are transcribed into the RNA molecules.
Afterwards, ribosome translates the RNA sequence into a protein that is based on the codon table.
Proteins are then folded into their natural 3D structure and prepared to realize their corresponding functions.

Despite of central dogma of molecular biology needs advanced compartments as ribosome for ensuring the flow of genetic material, not all organisms do have these compartments.
Viruses tends to exploit hosts mechanisms for expression of viral genes and replication of themselves.

We based our work on the central dogma of molecular biology and on the assumption that genes will be useful indicators when predicting the ability to infect certain hosts.

\begin{figure}[htp]
\includegraphics[width=\linewidth]{./images/central_dogma.png}
\centering
\cite{central_dogma}
\caption[Central dogma of molecular biology]{Central dogma of molecular biology,
arrows show the flow of information between biological molecules, full lines symbolizing usual flow in living organisms and dashed line symbolizing flow of information used by some primitive organisms (viruses)}
\end{figure}
% upravene end

\section{Bacteriophages}
\emph{Bacteriophages} (phages) belong to the group of viruses, which have the capability to infect and replicate within bacterial hosts.
It is estimated that they are the most abundant group of entities on Earth with their estimated count around $10^{31}$\cite{phage}.
In comparison, it is estimated that the count of all bacteria on planet Earth is around $10^{30}$.
They are one of the most highly diverse group in the biosphere, with their genomes spanning from few genes to as many as hundreds of genes.
We can find them in every place on earth, where bacteria are able to live, even inside our bodies.
It is believed that one of the most saturated location of their occurrence is sea, with  $9\cdot 10^8$ virions per milliliter of seawater at the surface and 70\% of bacteria infected \cite{virioplankton}.
% upravene start
Thanks to the high abundance of bacteriophages, their impact on shaping of the environment is outstandingly significant, reducing considerable amount of bacteria.
% upravene end

% + circulating carbon?
% + they contribute to horizontal gene transfer via transduction and transformation

\subsection{Taxonomical classification}
Classification of bacteriophages is difficult mainly due to their high diversity and genome mosaicism. \cite{phagetax}
Consequently, there exists no universal marker similar to universal markers in bacteria, according to which we will be able to classify bacteriophages based on their genetic information.
This is because no genes are strongly conserved within all bacteriophages.
Despite of these facts, there is taxonomical classification of phages.
This classification was created by International Committee on Taxonomy of Viruses (ICTV) and it categorizes each phage according to its morphology and nucleic acid.
This taxonomy recognizes nineteen different families of phages.
Although this taxonomy is currently in use, many biologists feel it is outdated and in need of revision. \cite{phagetax}

\subsection{Structure of a typical bacteriophage}
Given the high variety of bacteriophages, they come in a lot of different sizes and shapes.
Each bacteriophage consists of genetic information in form of DNA or RNA and capsid, a protein coat usually composited from higher number of identical protein units \cite{phage_coat}.
We will describe the most typical form of tailed bacteriophages, which is abundantly found in nature.

\begin{figure}[h]
\includegraphics[width=0.5\textwidth]{./images/phage.png}
\centering
% upravene start
\cite{phage_pic}
\caption[Structure of a typical bacteriophage]{Structure of a typical bacteriophage, it is composited by head, in which the genetic material is stored and tail, through which the genetic material is passed during the infection 
% upravene end
}
\label{fig:phage}
\end{figure}


As we can see in the Figure \ref{fig:phage}, this bacteriophages body consists of a head and a tail.
These parts are created from proteins and in addition, the head contains DNA or RNA of the bacteriophage genome.
The tail is used to attach and to inject the genetic code into bacteria.
At the end of the tail, there are proteins, which are able to bind to specific receptors on the surface of a bacteria.
Thanks to this mechanism, bacteriophages tend to have a high specificity in selection of their prey.

\subsection{Life cycle of bacteriophages}
Generally, there are two strategies that bacteriophages use to secure their survival and replication; the \emph{lysogenic cycle} and the \emph{lytic cycle} \cite{cycles}.
Viruses tend to use these strategies in different proportions, but usually prefer to choose one of them.
The lysogenic cycle results in incorporations of phage genetic information into host DNA or creation of a circular replicon in the bacterial cytoplasm.
By this approach, the genetic material of a phage inside the host, called prophage, is duplicated together with the host genome and after cell division, both daughter cells contain DNA of the bacteriophage.
In dependence on the following events, viral DNA can be released from hosts genome and start proliferation of new phages via the lytic cycle.
The lytic cycle is characterized by the lysis of bacterial cells membrane and their subsequent death.
It starts by injecting bacteriophage genome into bacteria.
After this step virus is not incorporated, but it compromises bacterial translation apparatus to produce more viruses.
Once enough virions have been produced, special viral proteins dissolve the bacterial cell and virions are released into surrounding space.

\subsection{Potential usage of bacteriophages}
Their ability to kill or alter behavior of highly specific strains of bacteria makes bacteriophages a valuable target for research.
Humankind tried to harness the power of phages since their discovery in the year 1917.
This discovery is attributed to a French-Canadian microbiologist Félix d'Hérelle, who experimented extensively with phages and introduced the concept of \emph{phage therapy}\cite{phages_in_nature}.
Unfortunately, after the discovery of antibiotics, research in this field suffered from insufficient funding and the development was significantly slowed down.\\
Nowadays, when humanity face the threat of multiresistant bacteria and phages could bring a new methods into the battle against pathogenic bacteria, the research in this field begins to flourish.
We can see the potential of their use in the field of food industry as the substances that are prolonging life and improve the quality of our food \cite{foodphage}.
In the field of pharmacy, they could be used as medicines for bacterial infections \cite{drugphage}.
Their potential of use is significant where we encounter the need to control the lives of microbial communities.
Moreover, due to their high specificity to a particular strain of bacteria, these new practices will likely be without adverse effects on natural microbiome.

% \section{Phage effectiveness}
% documented case of use from india
% georgia - cocktails
% bacterial mechanisms to prevent

\subsection{Host identification}
% how to find them biologically?
In our work we focus on prediction of phage hosts from genomic sequence.
We based our method on the assumption that bacteriophages with similar set of genes will infect the same bacterial host.
The decision to create host predicting models was made based on its straightforward use in searching for phages suitable for phage therapy.

Attempts to classify bacteriophages from a genomic sequence were already made.
In the study "The Phage Proteomic Tree: a Genome-Based Taxonomy for Phage"\cite{phage} from year 2002, researchers performed analysis based on genomes resulting in phylogenetic tree compatible with ICTV system.
In their work, they proved phages indeed do not contain any universal genetic marker, which could be used as sequence for classification.
They also showed that classification based on whole proteome of phage is a reasonable approach.
In 2016 a similar tool, called HostPhinder, for predicting hosts from genomic sequences appeared \cite{hostphinder}.
Their approach to classification was through rate of similar k-mers, short sequences of nucleotides of length $k$.
With this approach they achieved the results of 81\% of correctly predicted genera.
In our work we tried a different approach, where we looked for similar genes instead of rate of similar k-mers.

% +https://academic.oup.com/bioinformatics/article/33/19/3113/3964377
% +virhostmatcher
